### DISCUSSION TITLE
Getting error in automation in the PerformUpkeep function

### DISCUSSION THREAD

**Discussion by heheboii11** [Cyfrin/foundry-full-course-cu#1388] [Q&A] [answered]
<img width="1822" alt="Screenshot 2024-02-07 at 2 04 44 AM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/132984428/4815e156-6047-4548-9c80-33e0e0d2a923">
Getting this error I have even tried (bool Upkeep, ) = CheckUpkeep(bytes(""));
even this is throwing me an error

**Comment by heheboii11** [comment]
Invalid type for argument in function call. Invalid implicit conversion from literal_string "" to bytes calldata requested.  this is error

**Comment by Niferu** [comment]
Hello @heheboii11 

Please link your repository or give us full code of your `lottery.sol`

**Comment by heheboii11** [comment]
// Layout of Contract:
// version
// imports
// errors
// interfaces, libraries, contracts
// Type declarations
// State variables
// Events
// Modifiers
// Functions

// Layout of Functions:
// constructor
// receive function (if exists)
// fallback function (if exists)
// external
// public
// internal
// private
// internal & private view & pure functions
// external & public view & pure functions

//SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;
import {VRFCoordinatorV2Interface} from "lib/chainlink-brownie-contracts/contracts/src/v0.8/interfaces/VRFCoordinatorV2Interface.sol";
import {VRFConsumerBaseV2} from "lib/chainlink-brownie-contracts/contracts/src/v0.8/VRFConsumerBaseV2.sol";

/**
 * @title A sample Raffle Contract
 * @author Karthik
 * @notice This contarct is for creating a sample raffle contract better than our fundme contract.
 * @dev Implements Chainlink VRFv2
 */

contract lottery is VRFConsumerBaseV2 {
    /** errors */
    error Raffle__Notennougheth();
    error Raffle__NotTransferred();
    error Raffle__NotOpen();
    error Raffle__NotUpkeep(uint256 balance, uint256 length, RaffleState state);
    /** Type declaration */
    enum RaffleState {
        OPEN,
        CALCULATING
    }

    /** state variables */
    uint16 private constant REQUEST_CONFIRMATIONS = 3;
    uint32 private constant NUM_WORDS = 1;

    uint256 private immutable i_entrancefee;
    uint256 private immutable i_interval;
    VRFCoordinatorV2Interface private immutable i_vrfCoordinator;
    bytes32 private immutable i_keyhash;
    uint32 private immutable i_callbackgas;
    uint64 private immutable i_subscriptionId;
    address payable[] private s_players;
    uint256 private s_lastTimestamp;
    address private s_lastWinner;
    RaffleState private s_rafflestate;

    /** event */
    event Raffle_entered(address indexed player);
    event Raffle_PickedWinner(address indexed winner);

    constructor(
        uint256 entrancefee,
        uint256 interval,
        address vrfCordinator,
        bytes32 keyhash,
        uint64 subscriptionId,
        uint32 callbackgaslimit
    ) VRFConsumerBaseV2(vrfCordinator) {
        i_entrancefee = entrancefee;
        i_interval = interval;
        i_vrfCoordinator = VRFCoordinatorV2Interface(vrfCordinator);
        i_keyhash = keyhash;
        i_subscriptionId = subscriptionId;
        i_callbackgas = callbackgaslimit;
        s_lastTimestamp = block.timestamp;
        s_rafflestate = RaffleState.OPEN;
    }

    function enterraffle() external payable {
        if (s_rafflestate != RaffleState.OPEN) {
            revert Raffle__NotOpen();
        }
        if (msg.value >= i_entrancefee) {
            revert Raffle__Notennougheth();
        }
        s_players.push(payable(msg.sender));
        emit Raffle_entered(msg.sender);
    }

    function CheckUpKeep(
        bytes calldata /* checkdata */
    ) public view returns (bool Upkeep, bytes memory /* performdata */) {
        bool Timehaspasses = (block.timestamp - s_lastTimestamp) >= i_interval;
        bool Enoughbalance = address(this).balance > 0;
        bool Enoughpeople = s_players.length > 0;
        bool Raffleisopen = RaffleState.OPEN == s_rafflestate;
        Upkeep = (Timehaspasses &&
            Enoughbalance &&
            Enoughpeople &&
            Raffleisopen);
        return (Upkeep, "0x0");
    }

    function PerformUpKeep(bytes calldata /* performdata */) external {
        (bool Upkeep, ) = CheckUpKeep("");
        if (!Upkeep) {
            revert Raffle__NotUpkeep(
                address(this).balance,
                s_players.length,
                RaffleState(s_rafflestate)
            );
        }

        s_rafflestate = RaffleState.CALCULATING;
        uint256 requestId = i_vrfCoordinator.requestRandomWords(
            i_keyhash,
            i_subscriptionId,
            REQUEST_CONFIRMATIONS,
            i_callbackgas,
            NUM_WORDS
        );
    }

    function fulfillRandomWords(
        uint256 requestID,
        uint256[] memory randomWords
    ) internal override {
        uint256 IndexOfWinner = randomWords[0] % s_players.length;
        address payable winner = s_players[IndexOfWinner];
        s_lastWinner = winner;
        s_rafflestate = RaffleState.OPEN;
        s_players = new address payable[](0);
        s_lastTimestamp = block.timestamp;
        emit Raffle_PickedWinner(winner);
        (bool sent, ) = winner.call{value: address(this).balance}("");
        //require(sent, Raffle__NotTransferred);
        if (!sent) {
            revert Raffle__NotTransferred();
        }
    }

    /** Getter functions  */
    function get_entrancefee() external view returns (uint256) {
        return i_entrancefee;
    }
}

**Comment by Niferu** [comment, marked as answer]
Hello @heheboii11 

To fix your issue change below lane:

From:
```solidity
function CheckUpKeep(bytes calldata /* checkdata */) public view returns (bool Upkeep, bytes memory /* performdata */) {}
```

Into:
```solidity
function CheckUpKeep(bytes memory /* checkdata */) public view returns (bool Upkeep, bytes memory /* performdata */) {}
```

Just change `calldata` into `memory` as this is proper operator for this function
